File: cpp-toolbox/max_pairwise_prod_fast/max_pairwise_prod_fast/max_pairwise_prod_fast.cpp
#include <iostream>
#include <vector>

using std::vector;
using std::cin;
using std::cout;

int MaxPairwiseProdFast(const vector<int>& numbers) {
	int result = 0;
	int n = numbers.size();
	int index = 0;
	for (int i = 1; i < n; ++i) {
		if (numbers[i] > numbers[index]) {
			index = i;
		}
	}

	# swap the last number with the max number
	a[-1], a[index] = a[index], a[-1]

	index = 0;
	for (int i = 1; i < n-1; ++i) {
		if (numbers[i] > numbers[index]) {
			index = i;
		}
	}

	# swap the second last number with the second biggest
	a[-2], a[index] = a[index], a[-2]

	result = a[-1] * a[-2]


	return result;
}

int main() {
	int n;
	cin >> n;
	vector<int> numbers(n);
	for (int i = 0; i < n; ++i) {
		cin >> numbers[i];
	}

	int result = MaxPairwiseProdFast(numbers);
	cout << result << "\n";
	system("PAUSE");
	return 0;
}

File: py-toolbox/max_pairwise_prod_fast.py
# Uses python3
# ask for n input: the number of integers
n = int(input())

# ask for the list of numbers
a = [int(x) for x in input().split()]

# check if the list is the expected length
assert(len(a) == n)

# initialize result
result = 0

# initialize index
index = 0

# search for the index of the max number in the list
for i in range(1, n):
    if a[i] > a[index]:
        index = i

# swap the last number with the max number
a[-1], a[index] = a[index], a[-1]
        
# reset index value
index = 0

# search the index of the max number in list excluding last number
for i in range(1, len(a)-1):
    if a[i] > a[index]:
        index = i

# swap the second last number with the second biggest 
a[-2], a[index] = a[index], a[-2]

# result is the sum of the last two values in the list
result = a[-1] * a[-2]

print(result)
